### PR TITLE
feat: gemma3 sql to text recipe

### PR DESCRIPTION
**Pull request by ShubyM**
Adapting this from here: https://github.com/google-gemini/gemma-cookbook/blob/main/docs/core/huggingface_text_finetune_qlora.ipynb.

### CODE REVIEW

File: client/pyproject.toml [hunk 1]
```diff
@@ -27,4 +27,4 @@ functiongemma-demo = "functiongemma_sft:cli"
 
 [tool.setuptools]
 packages = []
-py-modules = ["cli", "sft", "rlvr", "functiongemma_sft", "piglatin_sft", "verify_ppo", "plot_metrics", "plot_logs"]
+py-modules = ["cli", "sft", "rlvr", "functiongemma_sft", "piglatin_sft", "texttosql_sft", "verify_ppo", "plot_metrics", "plot_logs"]
```
Comment (droot) [inline client/pyproject.toml:30]: we need to understand what this is ?
Probably right time to look into if our project.toml file is idiomatic for our setup. Can you pl. add an issue for this.

File: client/texttosql_sft.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    plt.close(fig)
+
+
+# ── Main ─────────────────────────────────────────────────────────────
+
+
+def run_training(config: Config, preset: str) -> dict[str, float]:
+    # 1. Create training client
+    try:
+        resp = requests.get(f"{config.base_url.rstrip('/')}/api/v1/get_server_capabilities", timeout=5.0)
```
Comment (droot) [inline client/texttosql_sft.py]: I am guessing tinker SDK must have a method for this. Can we simply use this.

Also, I would also suggest to move this out of core train function to keep the core function clutter free and focussed.

File: client/texttosql_sft.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        layer_name="gemma preset",
+    ),
+    "notebook": chz.Blueprint(Config).apply(
+        {"steps": 375, "batch_size": 8, "rank": 16, "learning_rate": 2e-4, "train_limit": 10_000, "eval_limit": 2_500, "eval_every": 125},
+        layer_name="notebook preset",
+    ),
+}
+
+
+# ── Helpers ──────────────────────────────────────────────────────────
```
Comment (droot) [inline client/texttosql_sft.py]: I think we should push the helpers down in the file.

File: client/texttosql_sft.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+def run_training(config: Config, preset: str) -> dict[str, float]:
+    # 1. Create training client
+    try:
+        resp = requests.get(f"{config.base_url.rstrip('/')}/api/v1/get_server_capabilities", timeout=5.0)
+        resp.raise_for_status()
+        caps = resp.json()
+    except Exception as exc:
+        raise RuntimeError(f"Open-RL server at {config.base_url} is not reachable. Start it with `make run-text-to-sql-server`.") from exc
+
+    print(f"Server ready at {config.base_url} | model={caps.get('default_model') or 'unset'}")
```
Comment (droot) [inline client/texttosql_sft.py]: I think we should start using logging module for these prints.

File: client/texttosql_sft.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            pos = 0
+        batch = [train_examples[order[i]] for i in range(pos, pos + batch_size)]
+        pos += batch_size
+
+        datums = [r["datum"] for r in batch]
+        active = sum(r["active_tokens"] for r in batch)
+
+        # Forward-backward + optimizer step
+        fwdbwd = trainer.forward_backward(datums, "cross_entropy").result()
+        trainer.optim_step(types.AdamParams(learning_rate=config.learning_rate, grad_clip_norm=config.grad_clip_norm)).result()
```
Comment (droot) [inline client/texttosql_sft.py]: we should be submitting both the operations forward_backward and optim_step at the same time so that they are processed together in the same cycle to keep the loop efficient.

File: client/texttosql_sft.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+        loss = float(fwdbwd.metrics.get("loss:sum", 0.0)) / max(1, active)
+        losses.append(loss)
+        print(f"[train] step={step:04d}/{config.steps} loss={loss:.4f}")
```
Comment (droot) [inline client/texttosql_sft.py]: I think we should follow tinker's cookbook practice here of writing the metrics in a jsonl file and also move our other recipes to this practice over time to keep it consistent.

One advantage of that is you can launch `make plot/or something` and point it to the metrics.jsonl and it can draw the graphs in real time.

File: server/src/engine.py [hunk 7]
```diff
@@ -57,7 +57,7 @@ def preload_base_model(self, base_model: str):
                     device_map=self.device
                 )
                 print(f"[EAGER INIT] {base_model} successfully seated in VRAM.")
-    def load_model(self, base_model: str, rank: int, model_id: str):
+    def load_model(self, base_model: str, rank: int, model_id: str, lora_config: Dict[str, Any] | None = None):
```
Comment (droot) [inline server/src/engine.py]: should we make `rank` also part of the lora_config ?

File: server/src/engine.py [hunk 8]
```diff
@@ -72,37 +72,52 @@ def load_model(self, base_model: str, rank: int, model_id: str):
                     torch_dtype=dtype,
                     device_map=self.device
                 )
+            config = lora_config or {}
+            train_attn = config.get("train_attn", True)
+            train_mlp = config.get("train_mlp", True)
+            train_unembed = config.get("train_unembed", True)
```
Comment (droot) [inline server/src/engine.py:79]: is there any doc that talks about these that we can refer to ?

File: server/src/engine.py [hunk 9]
```diff
@@ -72,37 +72,52 @@ def load_model(self, base_model: str, rank: int, model_id: str):
...
-                )
                 self.model.add_adapter(model_id, peft_config)
                 self.model.train()
             
         # Reset/initialize optimizer for this new adapter
         if model_id in self.optimizers:
             del self.optimizers[model_id]
             
-        self._auto_save_adapter(model_id)
+        if os.getenv("OPEN_RL_AUTO_SAVE_ADAPTERS", "0").lower() in {"1", "true", "yes", "on"}:
```
Comment (droot) [inline server/src/engine.py]: Is there any reason to make it configurable ? I am worried about too many knobs controlled through env variables.

File: server/src/engine.py [hunk 10]
```diff
@@ -478,13 +494,18 @@ async def clock_cycle_loop():
...
-                                    await asyncio.to_thread(engine.load_model, base_model, rank, m_id)
+                                    lora_config = r.get("lora_config", {})
+                                    await asyncio.to_thread(engine.load_model, base_model, rank, m_id, lora_config)
                                     await store.set_future(req_id, {
                                         "model_id": m_id,
                                         "is_lora": True,
                                         "lora_rank": rank,
                                         "type": "create_model"
                                     })
+                                elif req_type == "save_adapter":
```
Comment (droot) [inline server/src/engine.py]: is this checkpoint API implementation ?

File: Makefile [hunk 11]
```diff
@@ -41,6 +41,25 @@ run-pig-latin-gemma-server:
...
+	$(MAKE) run-text-to-sql-server TEXT_TO_SQL_SERVER_EXTRA=gpu TEXT_TO_SQL_SAMPLER_BACKEND=vllm
+
+run-text-to-sql-vllm:
+	cd server && UV_INDEX_URL="https://pypi.org/simple" CUDA_VISIBLE_DEVICES="$(VLLM_GPU)" VLLM_MODEL="$(TEXT_TO_SQL_BASE_MODEL)" uv run --extra gpu --extra vllm python -m src.vllm_worker
+
+TEXT_TO_SQL_PRESET ?= gemma
+
+run-text-to-sql-sft:
+	cd client && uv run --python 3.12 -i https://pypi.org/simple python -u texttosql_sft.py $(TEXT_TO_SQL_PRESET) base_url="http://127.0.0.1:9003" $(ARGS)
+
```
Comment (droot) [inline Makefile:62]: These makefile targets are getting out of hand :) We need to figure out a better way to scale these.

**Review by droot — COMMENTED** [review]
Thanks for working on it. I am looking forward to this example. I have minor comments mostly to make the code easier to follow.

File: client/texttosql_sft.py (hunk 2 above)
Comment (ShubyM) [reply client/texttosql_sft.py]: Yeah tinker has 3 types of clients: service, training and sampling. Any server information can fall under service client. See here for more: https://tinker-docs.thinkingmachines.ai/api-reference/serviceclient

File: server/src/engine.py (hunk 7 above)
Comment (ShubyM) [reply server/src/engine.py]: Yes we can match the tinker LoRA type: https://tinker-docs.thinkingmachines.ai/api-reference/types#loraconfig-objects

File: server/src/engine.py (hunk 8 above)
Comment (ShubyM) [reply server/src/engine.py:79]: They have this type: https://tinker-docs.thinkingmachines.ai/api-reference/types#loraconfig-objects

File: server/src/engine.py (hunk 9 above)
Comment (ShubyM) [reply server/src/engine.py]: Yea agreed this is not really needed. I had this in here for testing if it mattered for speed will remove

File: Makefile (hunk 11 above)
Comment (ShubyM) [reply Makefile:62]: Made a seperate issue for this

File: server/src/engine.py (hunk 10 above)
Comment (ShubyM) [reply server/src/engine.py]: Previously the behavior was that we would auto-save the adapter to disk after every training and optim step. Calling save just wrote additional meta data. This makes it so that save_weights and save_weights_for_sampler actually trigger a write to disk.

Comment (droot) [reply server/src/engine.py]: Is there a reason for remove auto-save after the optim_step?


auto_save was intentional. will try explaining here briefly, but may not be obvious at first because I learnt this during the parallel runs of GSM8K.

Without auto-save, each RL job schedules three requests for trainer worker in the queue (train, optim-step, save) and with auto-save, you are just left with two (train, optim-step). If you look at the RL training loop, it needs to wait for "save_weights" to finish before sampling. if there are multiple RL jobs, the round-robin scheduler picks up these items from the queue in round robin fashin, and if there are two jobs, save-weight step introduces dependencies on train/optim-step from other jobs to finish.

**Review by droot — COMMENTED** [review]
Everything looks good to go except the behavior change around save weights, I added an explanation why we have it like this.

**Review by droot — APPROVED** [review]
